Index: linux-3.3.8/arch/mips/ath79/prom.c
===================================================================
--- linux-3.3.8.orig/arch/mips/ath79/prom.c	2013-08-05 14:42:09.620683453 +0200
+++ linux-3.3.8/arch/mips/ath79/prom.c	2013-08-05 14:55:08.812702018 +0200
@@ -70,6 +70,35 @@
 	return ret;
 }
 
+#ifdef CONFIG_IMAGE_CMDLINE_HACK
+extern char __image_cmdline[];
+
+static int __init ath79_use_image_cmdline(void)
+{
+	char *p = __image_cmdline;
+	int replace = 0;
+
+	if (*p == '-') {
+		replace = 1;
+		p++;
+	}
+
+	if (*p == '\0')
+		return 0;
+
+	if (replace) {
+		strlcpy(arcs_cmdline, p, sizeof(arcs_cmdline));
+	} else {
+		strlcat(arcs_cmdline, " ", sizeof(arcs_cmdline));
+		strlcat(arcs_cmdline, p, sizeof(arcs_cmdline));
+	}
+
+	return 1;
+}
+#else
+static inline int ath79_use_image_cmdline(void) { return 0; }
+#endif
+
 static int __init ath79_prom_init_myloader(void)
 {
 	struct myloader_info *mylo;
@@ -98,6 +127,8 @@
 
 	ath79_prom_append_cmdline("ethaddr", mac_buf);
 
+	ath79_use_image_cmdline();
+
 	return 1;
 }
 
@@ -113,6 +144,9 @@
 			strlcat(arcs_cmdline, " ", sizeof(arcs_cmdline));
 			strlcat(arcs_cmdline, argv[i], sizeof(arcs_cmdline));
 		}
+
+	if (ath79_use_image_cmdline())
+		return;
 }
 
 void __init prom_init(void)
