Index: linux-3.3.8/arch/mips/ath79/Kconfig
===================================================================
--- linux-3.3.8.orig/arch/mips/ath79/Kconfig	2013-06-03 15:34:06.751323214 +0800
+++ linux-3.3.8/arch/mips/ath79/Kconfig	2013-06-03 15:34:20.859547187 +0800
@@ -71,6 +71,19 @@
 	  Say 'Y' here if you want your kernel to support the
 	  Atheros AP121 reference board.
 
+config ATH79_MACH_AP123
+	bool "Atheros AP123 reference board"
+	select SOC_AR934X
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_USB
+	select ATH79_DEV_WMAC
+	help
+	   Say 'Y' here if you want your kernel to support the
+	   Atheros AP123 reference board
+
 config ATH79_MACH_AP135
 	bool "Atheros AP135 reference board"
 	select SOC_QCA955X
Index: linux-3.3.8/arch/mips/ath79/Makefile
===================================================================
--- linux-3.3.8.orig/arch/mips/ath79/Makefile	2013-06-03 15:34:06.767323788 +0800
+++ linux-3.3.8/arch/mips/ath79/Makefile	2013-06-03 15:34:20.859547187 +0800
@@ -49,6 +49,7 @@
 obj-$(CONFIG_ATH79_MACH_ALL0315N)	+= mach-all0315n.o
 obj-$(CONFIG_ATH79_MACH_AP113)		+= mach-ap113.o
 obj-$(CONFIG_ATH79_MACH_AP121)		+= mach-ap121.o
+obj-$(CONFIG_ATH79_MACH_AP123)		+= mach-ap123.o
 obj-$(CONFIG_ATH79_MACH_AP135)		+= mach-ap135.o
 obj-$(CONFIG_ATH79_MACH_AP135_DUAL)	+= mach-ap135.o
 obj-$(CONFIG_ATH79_MACH_AP136)		+= mach-ap136.o
Index: linux-3.3.8/arch/mips/ath79/machtypes.h
===================================================================
--- linux-3.3.8.orig/arch/mips/ath79/machtypes.h	2013-06-03 15:34:06.787323992 +0800
+++ linux-3.3.8/arch/mips/ath79/machtypes.h	2013-06-03 15:34:20.863547306 +0800
@@ -25,6 +25,7 @@
 	ATH79_MACH_AP113,		/* Atheros AP113 reference board */
 	ATH79_MACH_AP121,		/* Atheros AP121 reference board */
 	ATH79_MACH_AP121_MINI,		/* Atheros AP121-MINI reference board */
+	ATH79_MACH_AP123,		/* Atheros AP123 reference board */
 	ATH79_MACH_AP135,		/* Atheros AP135 reference board */
 	ATH79_MACH_AP135_DUAL,		/* Atheros AP135 Dual reference board */
 	ATH79_MACH_AP136,		/* Atheros AP136 reference board */
