--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -746,6 +746,19 @@ config ATH79_MACH_NBG460N
 	select ATH79_DEV_M25P80
 	select ATH79_DEV_WMAC
 
+config ATH79_MACH_FON_2415
+	bool "FONERA 2415"
+	select ATH79_DEV_AUDIO
+	select SOC_AR934X
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_USB
+	select ATH79_DEV_WMAC
+	help
+	  Say 'Y' here if you want your kernel to support the
+	  Fonera Hub/Juke/Gramofon/... FON2415 board.
 endmenu
 
 config SOC_AR71XX
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -116,4 +116,5 @@ obj-$(CONFIG_ATH79_MACH_WZR_HP_G300NH2)
 obj-$(CONFIG_ATH79_MACH_WZR_HP_AG300H)	+= mach-wzr-hp-ag300h.o
 obj-$(CONFIG_ATH79_MACH_WZR_HP_G450H)	+= mach-wzr-hp-g450h.o
 obj-$(CONFIG_ATH79_MACH_ZCN_1523H)	+= mach-zcn-1523h.o
+obj-$(CONFIG_ATH79_MACH_FON_2415)	+= mach-fon2415.o
 
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -125,6 +125,7 @@ enum ath79_mach_type {
 	ATH79_MACH_WZR_HP_G450H,	/* Buffalo WZR-HP-G450H */
 	ATH79_MACH_ZCN_1523H_2,		/* Zcomax ZCN-1523H-2-xx */
 	ATH79_MACH_ZCN_1523H_5,		/* Zcomax ZCN-1523H-5-xx */
+	ATH79_MACH_FON2415,		/* FON 2415 */
 };
 
 #endif /* _ATH79_MACHTYPE_H */
