--- a/arch/mips/ath79/irq.c
+++ b/arch/mips/ath79/irq.c
@@ -27,27 +27,6 @@
 static void (*ath79_ip2_handler)(void);
 static void (*ath79_ip3_handler)(void);
 
-static void ath79_misc_irq_handler(unsigned int irq, struct irq_desc *desc)
-{
-	void __iomem *base = ath79_reset_base;
-	u32 pending;
-
-	pending = __raw_readl(base + AR71XX_RESET_REG_MISC_INT_STATUS) &
-		  __raw_readl(base + AR71XX_RESET_REG_MISC_INT_ENABLE);
-
-	if (!pending) {
-		spurious_interrupt();
-		return;
-	}
-
-	while (pending) {
-		int bit = __ffs(pending);
-
-		generic_handle_irq(ATH79_MISC_IRQ(bit));
-		pending &= ~BIT(bit);
-	}
-}
-
 static void ar71xx_misc_irq_unmask(struct irq_data *d)
 {
 	unsigned int irq = d->irq - ATH79_MISC_IRQ_BASE;
@@ -93,6 +72,36 @@ static struct irq_chip ath79_misc_irq_ch
 	.irq_mask	= ar71xx_misc_irq_mask,
 };
 
+static void ath79_misc_irq_handler(unsigned int irq, struct irq_desc *desc)
+{
+	void __iomem *base = ath79_reset_base;
+	u32 pending;
+
+	pending = __raw_readl(base + AR71XX_RESET_REG_MISC_INT_STATUS) &
+		  __raw_readl(base + AR71XX_RESET_REG_MISC_INT_ENABLE);
+
+	if (!pending) {
+		spurious_interrupt();
+		return;
+	}
+
+	while (pending) {
+		int bit = __ffs(pending);
+
+		generic_handle_irq(ATH79_MISC_IRQ(bit));
+
+		/* Interrupts must be cleared here since some of them require special
+		 * conditions (reading specific registers, resetting counters, etc.)
+		 * performed by generic_handle_irq in order to be properly cleared */
+		if(ath79_misc_irq_chip.irq_eoi) {
+			struct irq_data _irq_data;
+			_irq_data.irq = ATH79_MISC_IRQ(bit);
+			ath79_misc_irq_chip.irq_eoi(&_irq_data);
+		}
+		pending &= ~BIT(bit);
+	}
+}
+
 static void __init ath79_misc_irq_init(void)
 {
 	void __iomem *base = ath79_reset_base;
@@ -106,9 +115,10 @@ static void __init ath79_misc_irq_init(v
 	else if (soc_is_ar724x() ||
 		 soc_is_ar933x() ||
 		 soc_is_ar934x() ||
-		 soc_is_qca955x())
+		 soc_is_qca955x()) {
 		ath79_misc_irq_chip.irq_ack = ar724x_misc_irq_ack;
-	else
+		ath79_misc_irq_chip.irq_eoi = ar724x_misc_irq_ack;
+	} else
 		BUG();
 
 	for (i = ATH79_MISC_IRQ_BASE;
--- a/sound/soc/ath79/ath79-mbox.c
+++ b/sound/soc/ath79/ath79-mbox.c
@@ -71,10 +71,8 @@ void ath79_mbox_interrupt_enable(u32 mas
 void ath79_mbox_interrupt_ack(u32 mask)
 {
 	ath79_dma_wr(AR934X_DMA_REG_MBOX_INT_STATUS, mask);
-	ath79_reset_wr(AR71XX_RESET_REG_MISC_INT_STATUS, ~(MISC_INT_DMA));
 	/* Flush these two registers */
 	ath79_dma_rr(AR934X_DMA_REG_MBOX_INT_STATUS);
-	ath79_reset_rr(AR71XX_RESET_REG_MISC_INT_STATUS);
 }
 
 void ath79_mbox_dma_start(struct ath79_pcm_rt_priv *rtpriv)
